keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,510 @@
+
+## Design Details
+
+We are proposing to modify the defaulting generator to also generate new
+defaulting functions, which will use the marker to generate the default
+and apply them to the fields directly if the fields are set to their
+zero-value. These new defaulting functions will be called right before
+the existing defaulting functions.
+
+Because we also want server-side apply to be aware of these defaults,
+the OpenAPI generator will also use this information to fill the
+defaults fields in the OpenAPI. Then, Server-side apply will be able to use
+that additional information to default associative keys as needed.
+
+In other words, the plan would be the following:
+1. Update defaults generator to parse the `+default` marker and generate
+the additional defaulting functions, automatically call them prior to
+the existing defaulting functions
+2. Update kube-openapi to parse the `+default` marker and include it in
+the OpenAPI definition.
+3. Implement the new CRD semantics that clears `null` values from
+objects when the `x-kubernetes-remove-nulls` flag is present and true at
+the top-level (or in the CRD).
+4. Add additional "default" field to sigs.k8s.io/structured-merge-diff
+internal schema, and parse that field from the OpenAPI, use it when the
+map key is missing. The code for this already exists in a
+[PR](https://github.com/kubernetes-sigs/structured-merge-diff/pull/83#issuecomment-498410353)
+5. Update some of the existing built-in APIs to specify their default
+when they are strictly declarative, probably focusing first on defaults
+that are part of the associative keys (e.g. `protocol: TCP`).
+
+Note that this is not changing how built-in defaults work, but is merely
+adding an extra-layer on top of that which makes type authors life easier.
+
+### Examples
+
+#### Non-pointer structs
+
+Given the following type, defined either in Kubebuilder or as a built-in type:
+```golang
+type Root struct {
+  Entry SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      default: {}
+      type: object
+      properties:
+        name:
+          type: string
+          default: default-name
+        number:
+          type: integer
+          default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
```

Comment:
missing `default('{entry: {name:"",number:0}}')`

Reply:
Oh very good point, I want this nuance to be very clear.